src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.Analyzer.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns
+{
+    internal sealed partial class CSharpUseRecursivePatternsCodeRefactoringProvider
+    {
+        private sealed class Analyzer : CSharpSyntaxVisitor<AnalyzedNode>
+        {
+            private readonly SemanticModel _semanticModel;
+
+            private Analyzer(SemanticModel semanticModel)
+            {
+                _semanticModel = semanticModel;
+            }
+
+            public static AnalyzedNode Analyze(SyntaxNode node, SemanticModel semanticModel)
+            {
+                return new Analyzer(semanticModel).Visit(node);
+            }
+
+            public override AnalyzedNode VisitCasePatternSwitchLabel(CasePatternSwitchLabelSyntax node)
+            {
+                // Attempt to combine the pattern on the left and the condition of the when-clause.
+                if (node.WhenClause is var whenClause && whenClause != null)
+                {
+                    return new Conjunction(Visit(node.Pattern), Visit(whenClause.Condition));
+                }
+
+                return null;
+            }
+
+            public override AnalyzedNode VisitBinaryExpression(BinaryExpressionSyntax node)
+            {
+                var left = node.Left;
+                var right = node.Right;
+                switch (node.Kind())
+                {
+                    // Yield a pattern-match for an equality comparison if either side is a constant.
```

Comment:
these comments aren't great.  They're just descirbing what the code is doing.  What i would much rather see is an example of what the code would have started with and what it is being transformed into. (for all these cases).

Reply:
Yeah, I think I should just give examples, commenting a gear in a machine is not helpful.

Reply:
Note: when building trees, i really do like a: here's an example input, and here's how it is translated here.  

It *really* helps for understanding, checking, and building an intuition.  so i wouldn't mind both high level docs, as well as low level translation docs